File: go/codec.go
package wgrok

import (
	"bytes"
	"compress/gzip"
	"encoding/base64"
	"fmt"
	"io"
)

// Compress gzip-compresses data and returns a base64 string (no prefix).
func Compress(data string) (string, error) {
	var buf bytes.Buffer
	w := gzip.NewWriter(&buf)
	if _, err := w.Write([]byte(data)); err != nil {
		return "", fmt.Errorf("gzip write: %w", err)
	}
	if err := w.Close(); err != nil {
		return "", fmt.Errorf("gzip close: %w", err)
	}
	b64 := base64.StdEncoding.EncodeToString(buf.Bytes())
	return b64, nil
}

// Decompress attempts to base64-decode and gzip-decompress data.
// If decompression fails, returns the data unchanged (passthrough).
func Decompress(data string) (string, error) {
	compressed, err := base64.StdEncoding.DecodeString(data)
	if err != nil {
		// Not valid base64, passthrough
		return data, nil
	}
	r, err := gzip.NewReader(bytes.NewReader(compressed))
	if err != nil {
		// Not valid gzip, passthrough
		return data, nil
	}
	defer r.Close()
	out, err := io.ReadAll(r)
	if err != nil {
		// Gzip read failed, passthrough
		return data, nil
	}
	return string(out), nil
}

// Chunk splits payload into raw chunks of at most maxSize bytes each.
// Returns just the raw chunk data (no N/T: prefix).
func Chunk(payload string, maxSize int) ([]string, error) {
	if maxSize <= 0 {
		return nil, fmt.Errorf("maxSize must be positive")
	}
	total := (len(payload) + maxSize - 1) / maxSize
	if total == 0 {
		total = 1
	}
	chunks := make([]string, 0, total)
	for i := 0; i < total; i++ {
		start := i * maxSize
		end := start + maxSize
		if end > len(payload) {
			end = len(payload)
		}
		chunkData := payload[start:end]
		chunks = append(chunks, chunkData)
	}
	return chunks, nil
}

File: python/src/wgrok/codec.py
"""Payload codec: gzip+base64 compression and chunking for large payloads.

Pure data functions with no markers or prefixes.
"""

from __future__ import annotations

import base64
import gzip
import os


def compress(data: str) -> str:
    """Gzip compress and base64 encode. Returns raw base64 string (no prefix)."""
    compressed = gzip.compress(data.encode("utf-8"))
    return base64.b64encode(compressed).decode("ascii")


def decompress(data: str) -> str:
    """Base64 decode and gunzip. Input is raw base64 string."""
    compressed = base64.b64decode(data)
    return gzip.decompress(compressed).decode("utf-8")


def chunk(payload: str, max_size: int) -> list[str]:
    """Split payload into chunks of at most max_size characters each.

    Returns raw chunk strings (no N/T: prefix).
    """
    if max_size <= 0:
        raise ValueError("max_size must be positive")
    total = (len(payload) + max_size - 1) // max_size
    if total == 0:
        total = 1
    chunks = []
    for i in range(total):
        start = i * max_size
        end = start + max_size
        chunk_data = payload[start:end]
        chunks.append(chunk_data)
    return chunks


def encrypt(data: str, key: bytes) -> str:
    """AES-256-GCM encrypt. Returns base64(iv + ciphertext + tag).

    Args:
        data: Plain text string to encrypt.
        key: 32-byte encryption key.

    Returns:
        Base64-encoded encrypted payload (IV || ciphertext || 16-byte tag).
    """
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as e:
        raise ImportError(
            "encryption requires the 'cryptography' package. Install it with: pip install 'wgrok[crypto]'"
        ) from e

    if len(key) != 32:
        raise ValueError(f"encryption key must be 32 bytes, got {len(key)}")

    aesgcm = AESGCM(key)
    iv = os.urandom(12)
    ct = aesgcm.encrypt(iv, data.encode("utf-8"), None)
    return base64.b64encode(iv + ct).decode("ascii")


def decrypt(data: str, key: bytes) -> str:
    """AES-256-GCM decrypt. Input is base64(iv + ciphertext + tag).

    Args:
        data: Base64-encoded encrypted payload.
        key: 32-byte decryption key.

    Returns:
        Decrypted plain text string.
    """
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    except ImportError as e:
        raise ImportError(
            "decryption requires the 'cryptography' package. Install it with: pip install 'wgrok[crypto]'"
        ) from e

    if len(key) != 32:
        raise ValueError(f"decryption key must be 32 bytes, got {len(key)}")

    raw = base64.b64decode(data)
    if len(raw) < 28:
        raise ValueError(f"encrypted payload too short: expected at least 28 bytes (12 IV + 16 tag), got {len(raw)}")
    iv, ct = raw[:12], raw[12:]
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(iv, ct, None).decode("utf-8")
